CodingChallenge/MusicSearch/MusicSearchViewController.swift

```diff
@@ -0,0 +1,136 @@
+    
+    // MARK: Setup
+    
+    private func setup() {
+        let viewController = self
+        let interactor = MusicSearchInteractor()
+        let presenter = MusicSearchPresenter()
+        let router = MusicSearchRouter()
+        viewController.interactor = interactor
+        viewController.router = router
+        interactor.presenter = presenter
+        presenter.viewController = viewController
+        router.viewController = viewController
+        router.dataStore = interactor
+    }
+    
+    // MARK: Routing
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == "ShowSong" {
+            let data = sender as? SongResponse
+            let vc = segue.destination as! SongDetailViewController
+            vc.albumURL = data?.artworkUrl100 ?? ""
+            vc.artist = data?.artistName ?? ""
+            vc.track = data?.trackName ?? ""
+        }
+    }
+    
+    // MARK: View lifecycle
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        resultsCollection?.dataSource = resultsDataSource
+        resultsCollection?.delegate = self
+        resultsCollection?.register(UINib.init(nibName: "SearchResultCell", bundle: nil), forCellWithReuseIdentifier: "SearchResultCell")
+        resultsCollection?.isHidden = true
+        cancelButton?.isEnabled = false
+        searchStatusLabel?.text = SearchDisplayResults.firstMessage.rawValue
+    }
+    
+    // MARK: Search Function
+    func performSearch() {
+        let request = MusicSearch.Search.Request(searchTerm: searchTextField?.text ?? "")
+        searchStatusLabel?.text = SearchDisplayResults.loadingMessage.rawValue
+        UIApplication.shared.isNetworkActivityIndicatorVisible = true
+        interactor?.performSearch(request: request)
+    }
+    
+    func displayResults(viewModel: MusicSearch.Search.ViewModel) {
+        UIApplication.shared.isNetworkActivityIndicatorVisible = false
+        
+        guard let result = viewModel.searchResult else {
+            searchStatusLabel?.text = SearchDisplayResults.noResults.rawValue
+            return
+        }
+        
+        searchStatusLabel?.isHidden = true
+        resultsCollection?.isHidden = false
+        resultsDataSource.results = result
+        resultsCollection?.reloadData()
+    }
+    
+    // MARK: Text Delegate Functions
+    func textFieldDidBeginEditing(_ textField: UITextField) {
+        cancelButton?.isEnabled = true
+    }
+    
+    // MARK: Protocol implementation
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        textField.resignFirstResponder()
+        performSearch()
+        
+        return true
+    }
+    
+    // MARK: Collection View Delegate
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        //Navigate to detail
+        performSegue(withIdentifier: "ShowSong", sender: resultsDataSource.results[indexPath.row])
```

Comment:
I dig using enums to track my segue identifiers. Code completion, so hot right now.